dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/dataelement/DataElementMergeService.java

```diff
@@ -0,0 +1,226 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.merge.dataelement;
+
+import com.google.common.collect.ImmutableList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.PostConstruct;
+import javax.persistence.EntityManager;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.dataelement.DataElementService;
+import org.hisp.dhis.feedback.ErrorCode;
+import org.hisp.dhis.feedback.ErrorMessage;
+import org.hisp.dhis.feedback.MergeReport;
+import org.hisp.dhis.merge.CommonMergeHandler;
+import org.hisp.dhis.merge.MergeHandler;
+import org.hisp.dhis.merge.MergeParams;
+import org.hisp.dhis.merge.MergeRequest;
+import org.hisp.dhis.merge.MergeService;
+import org.hisp.dhis.merge.MergeValidator;
+import org.hisp.dhis.merge.dataelement.handler.AnalyticalDataElementMergeHandler;
+import org.hisp.dhis.merge.dataelement.handler.DataDataElementMergeHandler;
+import org.hisp.dhis.merge.dataelement.handler.MetadataDataElementMergeHandler;
+import org.hisp.dhis.merge.dataelement.handler.TrackerDataElementMergeHandler;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Main class for a {@link org.hisp.dhis.dataelement.DataElement} merge.
+ *
+ * @author david mackessy
+ */
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class DataElementMergeService implements MergeService {
+
+  private final DataElementService dataElementService;
+  private final MetadataDataElementMergeHandler metadataMergeHandler;
+  private final DataDataElementMergeHandler dataDataElementMergeHandler;
+  private final AnalyticalDataElementMergeHandler analyticalMergeHandler;
+  private final TrackerDataElementMergeHandler trackerMergeHandler;
+  private final CommonMergeHandler commonMergeHandler;
+  private final MergeValidator validator;
+  private final DataElementMergeValidator dataElementMergeValidator;
+  private final EntityManager entityManager;
+  private ImmutableList<DataElementMergeHandler> metadataMergeHandlers;
+  private ImmutableList<MergeHandler> commonMergeHandlers;
+  private ImmutableList<DataElementDataMergeHandler> dataMergeHandlers;
+  private ImmutableList<DataElementAuditMergeHandler> auditMergeHandlers;
+
+  @Override
```

Comment:
Is it intentional that `validate()` method doesn't have `@Transactional(readOnly=true)` ?

Reply:
We could add it here as a wrapper on the internal read only transactions that are already marked with `@Transactional(readOnly=true)`? What do you think?
The UID validation goes through and uses [this method](@Transactional(readOnly=true)) which is marked read only.